### DISCUSSION TITLE
NativeAOT

### DISCUSSION THREAD

**Discussion by murshex** [ravendb/ravendb#17939] [General]
Can I use RavenDB in a .NET 8 console app compiled with NativeAOT?

**Comment by ayende** [comment]
I don't believe that we have tested that scenario, but from a client perspective, I think it _should_ work. There is a lot of dynamic stuff inside of RavenDB (from JSON serialization to dynamic code emitting at runtime), but the client code is far simpler.
Would love to hear your report on this.

**Reply by SbiCA** [reply]
@ayende  that's why asked about using STJ instead of Newtonsoft.Json a while ago https://github.com/ravendb/ravendb/discussions/15301 that seems to one of the preconditions in order to use the source generate Json serialization

**Reply by SbiCA** [reply]
I didn't get the time to look into it but maybe @c0nd3v  will. I'd expect to see some perf improvements as a side effect of this 😃

**Reply by ayende** [reply]
Please note that Newtonsoft is just one of several aspects on the client where we'll use dynamic code generation, reflection, etc. I would really like to understand where you are seeing issues.

**Reply by SbiCA** [reply]
I'm not seeing issue but I assume that's one big part of the work to make a consumer able to use native AOT with Raven, since at least in our codebase this is a considerable amount of work (customer converters, swagger generation etc) that all need to be ported in order to make use of System.Text.Json see https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0#minimal-apis-and-json-payloads. 

In addition as a consumer you'd need to make sure you're on a supported part of the framework https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0#aspnet-core-and-native-aot-compatibility which leaves in our case  `Minimal APIs` which is only partially supported. In a nutshell I think porting an existing project will be challenging but do-able if you start from scratch probably "easier"

**Reply by SbiCA** [reply]
@ayende  I gave it a try and realized as an outsider I would say probably not possible without spending at least a few days on it 🥹  to understand the abstractions you put into place for Blittable Reader/Writers. 

I really had a tough time to understand how I should write the adapter for 
https://github.com/ravendb/ravendb/blob/f9e1a5717e164f37a287107a9ce2831e50575049/src/Raven.Client/Json/Serialization/NewtonsoftJson/NewtonsoftJsonSerializationConventions.cs mainly because it's internal hence can't be use from the outside therefore even if I could re-use things (which I'm not sure if a good idea or practically possible). 

In addition I'm also not quite clear if there are implications with using it in a mixed mode e.g. clients using STJ and Newtonsoft at the same time.

I pushed my code here https://github.com/SbiCA/Rebus-NativeAot/pull/1/files (but be aware it doesn't compile).

Happy to help if you guys have further questions but I'd really love to see this (https://blog.martincostello.com/native-aot-make-dotnet-lambda-go-brr/) happening with raven in the mix and generally I think the source generator could be a pref benefit as well

**Reply by SbiCA** [reply]
@ayende  @ppekrol are you guys thinking about making this happen or what's your view points?

**Reply by ppekrol** [reply]
We would love to take a peek there, but at the moment we are working on wrapping up the 7.0 and 7.1 releases. Using `<IsAotCompatible>true</IsAotCompatible>` in the csproj yields 200+ warnings. We are very heavily dependent on reflection and dynamics in the Client API, not sure even what would be the scope of work here. Looks a bit like an epic adventure.

**Reply by murshex** [reply]
It’s probably not that difficult to make the Client API support NativeAOT. System.Text.Json fully supports it.

**Reply by ayende** [reply]
We are using Newtonsoft by default here.
In general, would love to have Native AOT and System.JSON integrated.

But we are looking at doing that only after the 7.1 time frame (which is probably 4 - 6 months away)

Would be very happy to collaborate / assist if you care to give it a shot

**Reply by JohnGalt1717** [reply]
How is this progressing? I.e. with System.Text.Json as an option coming soon hopefully this would work great.  I created a library that you might be able to crib too with System.Text.Json so that it generates contexts automatically. My version takes all DataContext tagged classes/records and automaticlally forces them into the JsonContext stuff instead of having to use attributes. You could do something similar I would think but without the Datacontext attribute for all known data classes which would then further improve the perf of RavenDb against System.Text.Json and support AOT better: https://github.com/JohnGalt1717/MinimalSerializers

Would love to see full AOT Support added with the System.Text.Json work and be able to bid farwell to Newtonsoft.

**Reply by ayende** [reply]
That is something that we'll be looking at toward the 8.0 release time frame. 
System.Text.Json is in POC stage, and given the likelihood that it would be a big change, we'll put it in a major version.
Probably ~6 months out.

Edit: Adding disclaimer about this being a plan in my head, something we want to do, not firm commitment.

**Reply by JohnGalt1717** [reply]
We have a big product likely going to production around late November. Would love to release with System.Text.Json at least so that we don't have data migration issues later to get this type of functionality.

**Reply by ayende** [reply]
You can look at the open PR we have for this. 
But I don't think we'll have this out in a proper form by Nov (3 months from now), to be honest.

Do note that compatibility is going to be our biggest hurdle, and that is going to take a lot of time, probably.
The good news on your end is that you won't have to deal with data migrations when we get this out. There may be stuff you need to do in code (like different attributes on properties, different customizations if you wrote your own serialization logic, etc). 
But your data model will not change nor will the documents on the database.
